### ISSUE TITLE
MetaMask crashes when one tx in a batch fails.

### ISSUE THREAD

**Issue by danfinlay** [MetaMask/metamask-extension#1092]
From Denis:

> hey guys, one issue I’ve seen batch requests, is that for example if the response doesn’t have id, it breaks the apps, it loses connection and says this

<img width="416" alt="captura de pantalla 2017-02-07 a las 22 42 57" src="https://cloud.githubusercontent.com/assets/542863/22715602/7f167a1a-ed46-11e6-8588-b5949d916a5b.png">

> it happens when one of the requests of the batch have an error

**Comment by flyswatter** [comment]
Submitted by @denisgranha

**Comment by flyswatter** [comment]
Duplicate of #1092

**Comment by Zanibas** [comment]
One issue, related to the web3 API
```
Batch requests

Batch requests allow queuing up requests and processing them at once.

Note Batch requests are not faster! In fact making many requests at once will in some cases be faster, as requests are processed asynchronous. Batch requests are mainly useful to ensure the serial processing of requests.
```
Note on *serial processing*

Then look at this corresponding code:
```
    if (Array.isArray(payload)) {
      // handle batch
      async.map(payload, self._handleAsync.bind(self), cb)
```
And then the documentation for async.map:
```
Note, that since this function applies the iteratee to each item in parallel, there is no guarantee that the iteratee functions will complete in order. However, the results array will be in the same order as the original coll.
```
I recommend that we use async.mapSeries instead to be more in-line with the spec.
```
mapSeries(coll, iteratee, callbackopt)
The same as map but runs only a single async operation at a time.
```

**Comment by flyswatter** [comment]
That looks like a good proposal to me.

**Comment by kumavis** [comment]
I don't think batch is supposed to be serial.

according to the JSON RPC 2.0 spec 

> The Server MAY process a batch rpc call as a set of concurrent tasks, processing them in any order and with any width of parallelism.

http://www.jsonrpc.org/specification

**Comment by kumavis** [comment]
anyways these notes are rather separate from the original issue

**Comment by Zanibas** [comment]
Awww man, I was writing tests for it too in provider-engine.
Should we raise an issue in some other repo to clarify wording in the web3 repo then?

**Comment by kumavis** [comment]
yes we should

**Comment by Zanibas** [comment]
Opened
https://github.com/ethereum/wiki/issues/411

I'll continue investigating the original issue as well.

**Comment by bdresser** [comment]
closing due to inactivity – if this is still an issue for you, let us know
